cmasher/utils.py

```diff
@@ -233,6 +239,115 @@ def _get_cmap_perceptual_rank(
 
 
 # %% FUNCTIONS
+# This function combines multiple colormaps at given nodes
+def combine_cmaps(
+    *cmaps: Union[Colormap, str],
+    nodes: Optional[Union[list[float], np.ndarray]] = None,
+    n_rgb_levels: int = 256,
+    combined_cmap_name: str = "combined_cmap",
+) -> LinearSegmentedColormap:
+    """Create a composite matplotlib colormap by combining multiple colormaps.
+
+    Parameters
+    ----------
+        *cmaps: Colormap or colormap name (str) to be combined.
+        nodes: list or numpy array of nodes (float). Defaults: equal divisions.
+            The blending points between colormaps, in the range [0, 1].
+        n_rgb_levels: int. Defaults: 256.
+            Number of RGB levels for each colormap segment.
+        combined_cmap_name: str. Defaults: "combined_cmap".
+            name of the combined Colormap.
+
+    Returns
+    -------
+        Colormap: The composite colormap.
+
+    Raises
+    ------
+        TypeError: If the list contains mixed datatypes or invalid
+            colormap names.
+        ValueError: If the cmaps contain only one single colormap,
+            or if the number of nodes is not one less than the number
+            of colormaps, or if the nodes do not contain incrementing values
+            between 0.0 and 1.0.
+
+    Note
+    ----
+        The colormaps are combined from low value to high value end.
+
+    References
+    ----------
+        - https://stackoverflow.com/questions/31051488/combining-two-matplotlib-colormaps/31052741#31052741
+
+    Examples
+    --------
+        Using predefined colormap names::
+        >>> custom_cmap_1 = combine_cmaps(
+            ["ocean", "prism", "coolwarm"], nodes=[0.2, 0.75]
+        )
+
+        Using Colormap objects::
+        >>> cmap_0 = plt.get_cmap("Blues")
+        >>> cmap_1 = plt.get_cmap("Oranges")
+        >>> cmap_2 = plt.get_cmap("Greens")
+        >>> custom_cmap_2 = combine_cmaps([cmap_0, cmap_1, cmap_2])
+
+    """
+    # Check colormap datatype and convert to list[Colormap]
+    if len(cmaps) <= 1:
+        raise ValueError("Expected at least two colormaps to combine.")
+    for cm in cmaps:
+        if not isinstance(cm, (Colormap, str)):
+            raise TypeError(f"Unsupported colormap type: {type(cm)}.")
+    _cmaps: list[Colormap] = [
+        cm if isinstance(cm, Colormap) else mpl.colormaps[cm] for cm in cmaps
+    ]
+
+    # Generate default nodes for equal separation
+    if nodes is None:
+        nodes_arr = np.linspace(0, 1, len(_cmaps) + 1)[1:-1]
+    elif isinstance(nodes, list):
+        nodes_arr = np.array(nodes)
+    elif isinstance(nodes, np.ndarray):
+        nodes_arr = nodes
+    else:
+        raise TypeError(f"Unsupported nodes type: {type(nodes)}.")
```

Comment:
Let's avoid `np.append` and `np.insert`, which both unnecessary create copies. Instead, let's make exactly one copy every time with `np.concatenate`. I've also edited the length check to keep it consistent but I think the code would be clearer if it happened before the conversion to numpy array instead of after.
```suggestion
    if nodes is None:
         nodes_arr = np.linspace(0, 1, len(_cmaps) + 1)
    else:
        try:
            nodes_arr = np.concatenante([[0.0], nodes, [1.0]])
        except TypeError:
            raise TypeError(f"Unsupported nodes type: {type(nodes)}.") from None

    # Check nodes length
    if len(nodes_arr) != len(_cmaps) + 1:
        raise ValueError(
            "Number of nodes should be one less than the number of colormaps."
        )
```

Reply:
Thanks a lot for the suggestion, this logic looks cleaner. Just one thing, I noticed if I change it to:
```python
    # Generate default nodes for equal separation
    if nodes is None:
        nodes_arr = np.linspace(0, 1, len(_cmaps) + 1)
    else:
        try:
            # Insert the start/end points to make segmenting easier
            nodes_arr = np.concatenate([[0.0], nodes, [1.0]])
        except TypeError:
            raise TypeError(f"Unsupported nodes type: {type(nodes)}.") from None
```

I then got: 
```bash
ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 1 dimension(s) and the array at index 1 has 0 dimension(s)
```
for the invalid node type test (as the dimensionality of float `0.5` is 0, so a `ValueError` instead of `TypeError` is raised), I might still use explicit type checks to avoid this.

Reply:
I think the _test_ should be changed then. In my opinion `nodes=0.5` is a bit borderline, but it still makes sense if you're joining exactly 2 colormaps. Use a `str` a the invalid input for this test !

Reply:
Yes it makes sense, I would change it soon. Thanks!